### PR TITLE
Update ugorji/go/codec godep

### PR DESCRIPTION
**Pull request by sttts**
In order to pick-up https://github.com/ugorji/go/issues/119 and to get rid of the workaround at https://github.com/kubernetes/kubernetes/pull/36909/files#diff-a09eb061a0fb0ef3c9ef9d696f1ad0b4R426.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37557)

### CODE REVIEW

**Comment by liggitt** [comment]
we already bumped to a version that included that fix in https://github.com/kubernetes/kubernetes/pull/20536... did we switch branches or something?

**Comment by sttts** [comment]
Interesting. I see in master:

```
commit f1f1a805ed361a0e078bb537e4ea78cd37dcf065
Author: Ugorji Nwoke <ugorji@gmail.com>
Date:   Tue Oct 27 22:20:00 2015 -0400
```

**Comment by sttts** [comment]
Oh fun, https://github.com/kubernetes/kubernetes/pull/29399 updated etcd and pulled in an old ugorji/go lib.

**Comment by sttts** [comment]
@timothysc ^^

**Comment by sttts** [comment]
xref https://github.com/kubernetes/kubernetes/issues/36120

**Comment by wojtek-t** [comment]
@sttts - it may be hard to update just kubernetes and not etcd. I'm not a godeps expert, but it may be required to first update it in etcd, and only then update both ugorji and etcd godeps in k8s.

**Comment by sttts** [comment]
@wojtek-t I saw the old etcd update PRs and the discussion around parallel ugorji updates. Actually this dependency brought us back the broken ugorji version in the first place.

**Comment by liggitt** [comment]
looks like the version regression of ugorji also dropped https://github.com/ugorji/go/pull/137 (brought in in https://github.com/kubernetes/kubernetes/pull/20770) which means that API changes are now generating diffs with tens of thousands of lines of changes again

**Comment by liggitt** [comment]
increasing priority on this since we have a `nil/[]` serialization error currently

**Comment by wojtek-t** [comment]
@liggitt @sttts @lavalamp - should this be backported to 1.5 branch?

**Comment by sttts** [comment]
etcd is still using the same version we do (f1f1a805ed361a0e078bb537e4ea78cd37dcf065): https://github.com/gyuho/etcd/blob/master/glide.lock#L103

**Comment by liggitt** [comment]
> @liggitt @sttts @lavalamp - should this be backported to 1.5 branch?

I'd be concerned about any release without this

**Comment by timothysc** [comment]
@xiang90 @hongchaodeng this ugorji dep is transitive, could there will be a pr in flight soon that we should xref.

**Comment by sttts** [comment]
@timothysc test coverage for https://github.com/ugorji/go/issues/119 will be here (when the workaround is removed): https://github.com/kubernetes/kubernetes/pull/36909/files#diff-4eb9bb949f6a373a35016dac2f0d394aR70

**Comment by sttts** [comment]
Looks like the update ugorji version does not produce codecs for the base object (like `Job`) anymore, only `JobList`

**Comment by sttts** [comment]
Here we go: https://github.com/ugorji/go/commit/5d64d76d3ac4d5450009b0f5fa1530b790fccc71

All `runtime.Object` have `ObjectMeta` embedded. `ObjectMeta` is a self codec already, hence it is skipped by the generation (depending on the order of generation of the api group and checked in files 😱 ).

**Comment by liggitt** [comment]
> All runtime.Object have ObjectMeta embedded. ObjectMeta is a self codec already, hence it is skipped by the generation (depending on the order of generation of the api group and checked in files 😱 ).

Yeah, one of many reasons I'll be glad to see ugorji go. c.f. https://github.com/kubernetes/kubernetes/issues/36120#issuecomment-263375672

**Comment by sttts** [comment]
@liggitt ptal

File: hack/update-codecgen.sh [hunk 1]
```diff
@@ -144,7 +144,7 @@ for current in "${index[@]}"; do
   # We use '-d 1234' flag to have a deterministic output every time.
   # The constant was just randomly chosen.
   ###echo "DBG: running ${CODECGEN} -d 1234 -o ${base_generated_file} ${base_file}"
-  ${CODECGEN} -d 1234 -o "${base_generated_file}" "${base_file}"
+  ${CODECGEN} -d 1234 -o "${base_generated_file}" -nr 'ObjectMeta' "${base_file}"
```
Comment (liggitt) [inline hack/update-codecgen.sh]: won't skipping this cause problems?

File: cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go [hunk 2]
```diff
@@ -159,7 +159,13 @@ func (x *TestType) CodecEncodeSelf(e *codec1978.Encoder) {
 				z.EncSendContainerState(codecSelfer_containerArrayElem1234)
 				if yyq2[2] {
 					yy10 := &x.ObjectMeta
-					yy10.CodecEncodeSelf(e)
```
Comment (liggitt) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go]: falling back to reflective encoding on objectmeta is not what we want

Comment (sttts) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go]: indeed.

I wonder whether we can detect – via reflection in the codecgen tool – whether the self codec implementation comes from `ObjectMeta` or not.

Comment (liggitt) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go]: I don't see a way to detect promoted methods from embedded fields via reflection.

File: hack/update-codecgen.sh (hunk 1 above)
Comment (liggitt) [inline hack/update-codecgen.sh]: why isn't the dependency sorting in update-codecgen.sh sufficent for this already? It starts by removing the generated files, so ObjectMeta shouldn't have a generated encoding implementation during generation.

Comment (liggitt) [inline hack/update-codecgen.sh]: I think we should be able to drop this commit... the package generation order should be deterministic already

**Comment by liggitt** [comment]
LGTM once the ObjectMeta exclusion commit is removed

**Comment by liggitt** [comment]
I'd like a test that exercises the `nil`/`[]` distinction so we don't regress again, but I'd take it as a follow-up

File: hack/update-codecgen.sh (hunk 1 above)
Comment (sttts) [inline hack/update-codecgen.sh]: it is deterministic. But that's not the problem. The issue is that `ObjectMeta` (as a `api/v1` type) is processed first. All types embedding `ObjectMeta` get the promoted `CodecEncodeSelf` func and the ugorji generator skips the generation for that type.

File: cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go (hunk 2 above)
Comment (sttts) [inline cmd/libs/go2idl/client-gen/test_apis/testgroup/v1/types.generated.go]: Looks like there isn't any, by design.

File: hack/update-codecgen.sh (hunk 1 above)
Comment (sttts) [inline hack/update-codecgen.sh]: I think the cleanest solution (without changing the Selfer interface to have separate custom Selfer funcs and generated funcs) would be to teach codecgen to support multiple packages in one go.

Comment (sttts) [inline hack/update-codecgen.sh]: Making codecgen multi-package aware fixes the problem at hand. But it makes codec generation uncomposable, i.e. downstream has to remove and regenerate vendored codec files.

**Comment by sttts** [comment]
verify-godep complains about dependencies which are not touched in this PR :-/

**Comment by liggitt** [comment]
looks like https://github.com/kubernetes/kubernetes/pull/39448 removed uses of skydns without removing corresponding godeps

**Comment by sttts** [comment]
@bowei ^^

**Comment by liggitt** [comment]
godeps fixed in https://github.com/kubernetes/kubernetes/pull/39624
@k8s-bot verify test this

**Comment by sttts** [comment]
As follow-up: we need some tests for the regression. And in the verbs type we can remove the Selfer implementation.

**Comment by liggitt** [comment]
can you title the godep bump commit as `bump(ugoriji/go/codec): ded73eae5db7e7a0ef6f55aace87a2873c5d2b74`

**Comment by sttts** [comment]
@k8s-bot verify test this

**Comment by sttts** [comment]
@k8s-bot unit test this issue #39604

**Comment by sttts** [comment]
@k8s-bot unit test this issue #39604

**Comment by sttts** [comment]
@liggitt I added the tests and removed the Selfer impl. May I assume that your old lgtm is still valid?

**Comment by liggitt** [comment]
yes /lgtm

**Comment by deads2k** [comment]
@sttts  I'd like to hold this PR until after https://github.com/kubernetes/kubernetes/pull/39475.  I don't want to risk that pull for this bump.

**Comment by sttts** [comment]
@k8s-bot gci gke e2e test this

**Comment by sttts** [comment]
@k8s-bot cvm gke e2e test this

**Comment by deads2k** [comment]
@sttts released hold.  Looks like only one conflict.

**Comment by sttts** [comment]
Rebased after apimachinery PR merged.

**Comment by sttts** [comment]
@k8s-bot test this

File: hack/update-codecgen.sh [hunk 3]
```diff
@@ -31,7 +31,7 @@ generated_files=($(
       \( \
         -wholename './output' \
         -o -wholename './_output' \
-        -o -wholename './staging' \
```
Comment (sttts) [inline hack/update-codecgen.sh]: @smarterclayton you might want this. Will remove that here again as it is not yet necessary.

File: hack/verify-readonly-packages.sh [hunk 4]
```diff
@@ -45,7 +45,7 @@ find_files() {
 IFS=$'\n'
 conflicts=($(find_files | sed 's|/.readonly||' | while read dir; do
     dir=${dir#./}
-    if kube::util::has_changes_against_upstream_branch "${branch}" "^${dir}/[^/]*\$" "/.readonly\$|/BUILD$|/zz_generated" &>/dev/null; then
+    if kube::util::has_changes_against_upstream_branch "${branch}" "^${dir}/[^/]*\$" '/\.readonly$|/BUILD$|/zz_generated|/\.generated\.|\.proto$|\.pb\.go$' >/dev/null; then
```
Comment (deads2k) [inline hack/verify-readonly-packages.sh]: `.generated.` is for `types.generated.go`?

Comment (deads2k) [inline hack/verify-readonly-packages.sh]: and the other is generated.pb.go?

I might have listed the files explicitly, but I don't feel strongly.

**Comment by liggitt** [comment]
```
I0111 13:24:54.839] <         "//vendor:github.com/ugorji/go/codec",
I0111 13:24:54.848] 
I0111 13:24:54.848] Run ./hack/update-bazel.sh
I0111 13:24:54.850] FAILED   hack/make-rules/../../hack/verify-bazel.sh	
```

**Comment by sttts** [comment]
@k8s-bot bazel test this

**Pull request merged**